### PR TITLE
User/jason/182979 installintegration

### PR DESCRIPTION
**Pull request by ctediz**
Integrate the identity secret handling changes into the identity install script

### CODE REVIEW

File: Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 [hunk 1]
```diff
@@ -1982,30 +1982,77 @@ function Remove-FilePermissions
...
     return $encryptionCertificate.NotAfter -gt $today
 }
 
 function Remove-IdentityEncryptionCertificate {
     param(
-        [string] $encryptionCertificateThumbprint
+        [string] $encryptionCertificateThumbprint,
+        [string] $friendlyName = "Fabric Identity Signing Encryption Certificate"
     )
-    $cert = Get-Certificate $encryptionCertificateThumbprint
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1]: Fun note: Authorization uses this function as well, this causes the Authorization install.config to have a signing/encryption certificate in it's own install.config section.

Authorization will need to be updated to either not use this, or to grab the encryption certificate value from the common section of the install.config

File: Fabric.Identity.API/scripts/Install-Identity.ps1 [hunk 2]
```diff
@@ -71,14 +68,40 @@ $currentDirectory = $PSScriptRoot
...
-$iisUser = Get-IISAppPoolUser -credential $credential -appName $installSettings.appName -storedIisUser $installSettings.iisUser -installConfigPath $configStore.Path -scope $installSettingsScope
-Add-PermissionToPrivateKey $iisUser.UserName $selectedCerts.SigningCertificate read
 $appInsightsKey = Get-AppInsightsKey -appInsightsInstrumentationKey $installSettings.appInsightsInstrumentationKey -installConfigPath $configStore.Path -scope $installSettingsScope -quiet $quiet
 $sqlServerAddress = Get-SqlServerAddress -sqlServerAddress $commonInstallSettings.sqlServerAddress -installConfigPath $configStore.Path -quiet $quiet
 $identityDatabase = Get-IdentityDatabaseConnectionString -identityDbName $installSettings.identityDbName -sqlServerAddress $sqlServerAddress -installConfigPath $configStore.Path -quiet $quiet
 $metadataDatabase = Get-MetadataDatabaseConnectionString -metadataDbName $commonInstallSettings.metadataDbName -sqlServerAddress $sqlServerAddress -installConfigPath $configStore.Path -quiet $quiet
 
+# Switch to DosInstallUtilities later, testing with Fabric-Install-Utilities locally
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity.ps1]: Seems like a good section to move into its own function and add tests around

File: Fabric.Identity.API/scripts/Install-Identity.ps1 [hunk 3]
```diff
@@ -71,14 +68,40 @@ $currentDirectory = $PSScriptRoot
...
+    Remove-IdentityEncryptionCertificate -encryptionCertificateThumbprint $commonInstallSettings.encryptionCertificateThumbprint
+    $encryptionCertificate = New-IdentityEncryptionCertificate
+}
+
+# Wait until potentially creating new certificate to add permissions to private key
+$iisUser = Get-IISAppPoolUser -credential $credential -appName $installSettings.appName -storedIisUser $installSettings.iisUser -installConfigPath $configStore.Path -scope $installSettingsScope
+Add-PermissionToPrivateKey $iisUser.UserName $encryptionCertificate read
+
+# update install.config
+# Assumes both encryption and signing cert are the same certificate
```
Comment (ctediz) [inline Fabric.Identity.API/scripts/Install-Identity.ps1]: Important note here ⏫

Comment (ctediz) [reply Fabric.Identity.API/scripts/Install-Identity.ps1]: Currently, the only way to separate the encryption/signing certificate is to do so manually through the install.config.  
The Get-Certificates behavior assumed that these would be the same as well.
With this new implementation, everytime the new cert is generated, then both the encryption/signing certs are replaced.

Is this assumption still valid?

File: Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1 (hunk 1 above)
Comment (hckenmiller) [reply Fabric.Identity.API/scripts/Install-Identity-Utilities.psm1]: I like having the common stuff like encryption certificate in common section.

File: Fabric.Identity.API/scripts/Install-Identity.ps1 (hunk 3 above)
Comment (hckenmiller) [reply Fabric.Identity.API/scripts/Install-Identity.ps1]: I always remembered them showing the same value, and it seemed weird to me that they are the same but separated out.

File: Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+                $thumbPrint = "non-existing-thumbprint"
+                Mock -ModuleName Install-Identity-Utilities -CommandName Write-DosMessage -MockWith {}
+                Mock -ModuleName Install-Identity-Utilities -CommandName Remove-Item -MockWith {}
+
+                # Act
+                Remove-IdentityEncryptionCertificate -encryptionCertificateThumbprint $thumbPrint
+
+                # Assert
+                Assert-MockCalled -ModuleName Install-Identity-Utilities -CommandName Write-DosMessage -Times 1 -Exactly -Scope It
+                Assert-MockCalled -ModuleName Install-Identity-Utilities -CommandName Write-DosMessage -ParameterFilter { $Level -and $Level -eq "Information" -and $Message.StartsWith("Certificate with thumbprint") } -Times 1 -Exactly -Scope It
```
Comment (hckenmiller) [inline Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1]: Does the message also contain "not found"?

File: Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+        }
+
+        InModuleScope Install-Identity-Utilities {
+            It 'Should not attempt to remove when certificate is not the identity encryption certificate' {
+                # Arrange
+                $testCert = New-Object -TypeName System.Security.Cryptography.X509Certificates.x509Certificate2
+                $friendlyNameProperty = @{
+                    MemberType = [System.Management.Automation.PSMemberTypes]::NoteProperty
+                    Name = "FriendlyName"
+                    value = "Some Friendly Name"
```
Comment (hckenmiller) [inline Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1]: So the value in the friendlyNameProperty would have to be "Fabric Identity Signing Encryption Certificate" for it to be removed?

File: Fabric.Identity.API/scripts/Install-Identity.ps1 (hunk 2 above)
Comment (hckenmiller) [reply Fabric.Identity.API/scripts/Install-Identity.ps1]: yah.

File: Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1 (hunk 5 above)
Comment (ctediz) [reply Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1]: Ya, it defaults to "Fabric Identity Signing Encryption Certificate" to look for, but that can be passed in to be something else.

The idea was to not accidently remove the SSL certificate, or any certificate that we did not create ourselves.

File: Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1 (hunk 4 above)
Comment (ctediz) [reply Fabric.Identity.API/scripts/tests/Remove-IdentityEncryptionCertificate.tests.ps1]: Ya, i can make the message to take for be more explicit

**Pull request merged**